wled00/pin_manager.h

```diff
@@ -4,26 +4,103 @@
+  // #define USERMOD_ID_DHT                             // 0x0A // Usermod "usermod_dht.h" -- Statically allocates pins, not compatible with pinManager?
+  // #define USERMOD_ID_MODE_SORT                       // 0x0B // Usermod "usermod_v2_mode_sort.h" -- Does not allocate pins
+  // #define USERMOD_ID_VL53L0X                         // 0x0C // Usermod "usermod_vl53l0x_gestures.h" -- Uses "standard" I2C pins ... TODO -- enable shared I2C bus use
+  UM_MultiRelay        = USERMOD_ID_MULTI_RELAY,        // 0x0D // Usermod "usermod_multi_relay.h"
+  UM_AnimatedStaircase = USERMOD_ID_ANIMATED_STAIRCASE, // 0x0E // Usermod "Animated_Staircase.h"
+  // #define USERMOD_ID_RTC                             // 0x0F // Usermod "usermod_rtc.h" -- Uses "standard" I2C pins ... TODO -- enable shared I2C bus use
+  // #define USERMOD_ID_ELEKSTUBE_IPS                   // 0x10 // Usermod "usermod_elekstube_ips.h" -- Uses quite a few pins ... see Hardware.h and User_Setup.h
+  // #define USERMOD_ID_SN_PHOTORESISTOR                // 0x11 // Usermod "usermod_sn_photoresistor.h" -- Uses hard-coded pin (PHOTORESISTOR_PIN == A0), but could be easily updated to use pinManager
+};
+static_assert(0u == static_cast<uint8_t>(PinOwner::None), "PinOwner::None must be zero, so default array initialization works as expected");
+
+// Allocates a single pin, with an owner tag.
+// De-allocation requires the same owner tag (or override)
+#define ALLOCATE_PIN(gpio, output, tag)     pinManager._allocatePin(gpio, output, tag)
+// Allocates all the pins, or allocates none of the pins, with owner tag.
+// Provided to simplify error condition handling in clients 
+// using more than one pin, such as I2C, SPI, rotary encoders,
+// ethernet, etc..
+#define ALLOCATE_MULTIPLE_PINS(mptArray, arrayElementCount, tag) pinManager._allocateMultiplePins(mptArray, arrayElementCount, tag)
+// De-allocates a single pin
+#define DEALLOCATE_PIN(gpio, tag) pinManager._deallocatePin(gpio, tag)
+
 
 class PinManagerClass {
   private:
   #ifdef ESP8266
   uint8_t pinAlloc[3] = {0x00, 0x00, 0x00}; //24bit, 1 bit per pin, we use first 17bits
+  PinOwner ownerTag[17] = { PinOwner::None };
   #else
   uint8_t pinAlloc[5] = {0x00, 0x00, 0x00, 0x00, 0x00}; //40bit, 1 bit per pin, we use all bits
   uint8_t ledcAlloc[2] = {0x00, 0x00}; //16 LEDC channels
+  PinOwner ownerTag[40] = { PinOwner::None };
   #endif
 
   public:
-  void deallocatePin(byte gpio);
-  bool allocatePin(byte gpio, bool output = true);
-  bool isPinAllocated(byte gpio);
+  bool _deallocatePin(byte gpio, PinOwner tag);
+  bool _allocatePin(byte gpio, bool output, PinOwner tag);
+  bool _allocateMultiplePins(const managed_pin_type * mptArray, byte arrayElementCount, PinOwner tag );
+
+  #if !defined(ESP8266) // ESP8266 compiler doesn't understand deprecated attribute
+  [[deprecated("Replaced by macro ALLOCATE_PIN(gpio, output), for improved debugging")]]
```

Comment:
By marking the old API deprecated, it provides a backwards-compatible API surface, while making clear where older code should be updated.